### PR TITLE
Enable automatic re-bootstrapping of managed gardenlets if their client certificate is expired

### LINKED ISSUE #4687: New seed bootstrap tokens cannot be created for existing seeds
**How to categorize this issue?**

/area robustness ops-productivity
/kind bug
/priority 3

**What happened**:
The client certificate of one of the managedseed clusters has expired and it has not been renewed on purpose. 
Then, the parent gardenlet can create new bootstrap token and let the child seed use it to request new certificate. According to the code https://github.com/gardener/gardener/blob/981aa35da912171e88f13df2be201df5bd294185/pkg/gardenlet/controller/managedseed/managedseed_actuator.go#L687-L694, to trigger this action the secret from `gardenClientConnection.kubeconfigSecret` of the child seeds must not exist. I have deleted it, but then the parent gardenlet is failing to generate the token with error like
```
could not deploy gardenlet into shoot garden/foo: secrets "bootstrap-token-5111ca" is forbidden: User "gardener.cloud:system:seed:bar" cannot get resource "secrets" in API group "" in the namespace "kube-system": no relationship found between seed ''bar' and this object
```

This gardener landscape has the SeedAuthorizer enabled and it is indeed rejecting the request with
```
{"level":"info","ts":"2021-09-18T06:46:45.704Z","logger":"seedauthorizer","msg":"SEED DENY: 'bar' authorizer.AttributesRecord{User:(*user.DefaultInfo)(0xc001db1e40), Verb:\"get\", Namespace:\"kube-system\", APIGroup:\"\", APIVersion:\"v1\", Resource:\"secrets\", Subresource:\"\", Name:\"bootstrap-token-5111ca\", ResourceRequest:true, Path:\"\"}"} 
```

**What you expected to happen**:
A new bootstrap token to be successfully generated and gardenlet re-deployed.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:

- Gardener version: 981aa35da912171e88f13df2be201df5bd294185
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### LINKED ISSUE #4734: When gardenlet is using client certificate for the g API, it should log a message if it is still valid
**How to categorize this issue?**

/area ops-productivity quality
/kind enhancement
/priority 3

**What would you like to be added**:
The gardenlet to log a message with the expiration date of the client certificate for the garden cluster.

**Why is this needed**:
Currently, if gardenlet is started with expired client certificate, it will fail to discover the k8s version of the garden cluster, something like
```
Error: failed to get garden client: error creating new ClientSet for key "garden": error discovering kubernetes version: the server has asked for the client to provide credentials
```
Obviously, this error is pointing to issue with the client authentication, but it is not clear that actually the certificate has expired and needs to be renewed.

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security ops-productivity
/kind enhancement
/merge squash

**What this PR does / why we need it**:
This PR enables the automatic re-bootstrapping of managed gardenlets if their client certificate is expired.
This wasn't possible before since the `Seed{Authorizer,Restriction}` plugins weren't aware of the gardenlet's client certificate expiration timestamps and hence unable to take an educated decision whether re-bootstrapping is allowed or not.

To enable this, the idea is the following:

1. gardenlet's report their client certificate's expiration timestamp in the new `.status.clientCertificateExpirationTimestamp` field of their `Seed` resources
2. This allows the `Seed{Authorizer,Restriction}` plugins to decide whether the re-bootstrapping process is allowed or not.
3. The `ManagedSeed` controller in the (root) gardenlet will automatically delete the `gardenlet-kubeconfig` secret in the  seed in case the certificate has expired in order to trigger the re-bootstrapping process.
4. The seed lifecycle controller in GCM will automatically trigger respective `ManagedSeed`s for reconciliations if a gardenlet does no longer report its status and the client certificate has expired.

/cc @stoyanr @timebertt @vpnachev 

**Which issue(s) this PR fixes**:
Fixes #4687
Fixes #4734

**Release note**:

```feature operator
`gardenlet`s now report the expiration date of their client certificates in the `.status.clientCertificateExpirationTimestamp` field of their respective `Seed` resources. If they are managed by `ManagedSeed`s then this enables to re-bootstrap gardenlets in case their certificate expired and they weren't able to refresh it themselves. `gardener-controller-manager` will automatically trigger this process if necessary.
```

### CODE REVIEW

File: pkg/apis/core/types_seed.go [hunk 1]
```diff
@@ -104,6 +104,8 @@ type SeedStatus struct {
 	// Allocatable represents the resources of a seed that are available for scheduling.
 	// Defaults to Capacity.
 	Allocatable corev1.ResourceList
+	// ClientCertificateExpirationTimestamp is the Kubernetes version of the seed cluster.
+	ClientCertificateExpirationTimestamp *metav1.Time
```
Comment (stoyanr) [inline pkg/apis/core/types_seed.go]: The comment is doesn't match the field.

File: pkg/apis/core/v1alpha1/types_seed.go [hunk 2]
```diff
@@ -104,6 +104,9 @@ type SeedStatus struct {
 	// ClusterIdentity is the identity of the Seed cluster
 	// +optional
 	ClusterIdentity *string `json:"clusterIdentity,omitempty" protobuf:"bytes,5,opt,name=clusterIdentity"`
+	// ClientCertificateExpirationTimestamp is the Kubernetes version of the seed cluster.
+	// +optional
+	ClientCertificateExpirationTimestamp *metav1.Time `json:"clientCertificateExpirationTimestamp,omitempty" protobuf:"bytes,6,opt,name=clientCertificateExpirationTimestamp"`
```
Comment (stoyanr) [inline pkg/apis/core/v1alpha1/types_seed.go]: The comment is doesn't match the field.

File: pkg/apis/core/v1beta1/types_seed.go [hunk 3]
```diff
@@ -117,6 +117,9 @@ type SeedStatus struct {
 	// Defaults to Capacity.
 	// +optional
 	Allocatable corev1.ResourceList `json:"allocatable,omitempty" protobuf:"bytes,7,rep,name=allocatable"`
+	// ClientCertificateExpirationTimestamp is the Kubernetes version of the seed cluster.
+	// +optional
+	ClientCertificateExpirationTimestamp *metav1.Time `json:"clientCertificateExpirationTimestamp,omitempty" protobuf:"bytes,8,opt,name=clientCertificateExpirationTimestamp"`
```
Comment (stoyanr) [inline pkg/apis/core/v1beta1/types_seed.go]: The comment is doesn't match the field.

File: pkg/apis/core/types_seed.go (hunk 1 above)
Comment (stoyanr) [inline pkg/apis/core/types_seed.go]: Nit: The name of the field is a bit long. While this is not a big problem of course, perhaps it could be shortened to `ClientCertificateExpiration` or `ClientCertificateExpiresOn`? I think both these names already imply a `Time` type, so `Timestamp` adds to the length without adding much additional information. Of course if we rename it here, it should be renamed everywhere else for consistency.

File: pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_managedseed.go [hunk 4]
```diff
@@ -44,31 +42,11 @@ func (g *graph) setupManagedSeedWatch(ctx context.Context, informer cache.Inform
...
-			if err != nil {
-				return
-			}
-
-			if oldManagedSeed.Spec.Shoot.Name != newManagedSeed.Spec.Shoot.Name ||
-				!gardencorev1beta1helper.SeedBackupSecretRefEqual(oldSeedTemplate.Spec.Backup, newSeedTemplate.Spec.Backup) ||
-				!apiequality.Semantic.DeepEqual(oldSeedTemplate.Spec.SecretRef, newSeedTemplate.Spec.SecretRef) ||
-				!gardenletBootstrapModeEqual(oldManagedSeed.Spec.Gardenlet, newManagedSeed.Spec.Gardenlet) {
-				g.handleManagedSeedCreateOrUpdate(ctx, newManagedSeed)
-			}
```
Comment (stoyanr) [inline pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_managedseed.go:71]: Why was this code removed? It doesn't seem to be related to the rest of the change, did you realize it's not needed, or perhaps that the checks were not really necessary?

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go [hunk 5]
```diff
@@ -345,13 +346,18 @@ func (a *actuator) deployGardenlet(
 	gardenletConfig *configv1alpha1.GardenletConfiguration,
 	shoot *gardencorev1beta1.Shoot,
 ) error {
+	seed, err := a.getSeed(ctx, managedSeed)
+	if err != nil {
+		return err
+	}
+
```
Comment (stoyanr) [inline pkg/gardenlet/controller/managedseed/managedseed_actuator.go]: For consistency, I would put this code in `Reconcile` and pass the seed to this function as a parameter, as it's done in `deleteGardenlet`. (to maintain the symmetry between `deployGardenlet` and `deleteGardenlet`).

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go [hunk 6]
```diff
@@ -675,7 +680,17 @@ func ensureGardenletEnvironment(deployment *seedmanagementv1alpha1.GardenletDepl
...
+	ctx context.Context,
+	shootClient kubernetes.Interface,
+	gcc *configv1alpha1.GardenClientConnection,
+	managedSeed *seedmanagementv1alpha1.ManagedSeed,
+	seed *gardencorev1beta1.Seed,
+	bootstrap seedmanagementv1alpha1.Bootstrap,
+) (
+	string,
+	error,
+) {
```
Comment (stoyanr) [inline pkg/gardenlet/controller/managedseed/managedseed_actuator.go:694]: Nit: How about fewer lines of code?
```suggestion
) (string, error) {
```

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go [hunk 7]
```diff
@@ -684,13 +699,21 @@ func (a *actuator) prepareGardenClientConnectionWithBootstrap(ctx context.Contex
...
-	isAlreadyBootstrapped, err := isAlreadyBootstrapped(ctx, shootClient.Client(), gcc.KubeconfigSecret)
-	if err != nil {
-		return "", err
-	}
+	if seed != nil && seed.Status.ClientCertificateExpirationTimestamp != nil && seed.Status.ClientCertificateExpirationTimestamp.UTC().Before(time.Now().UTC()) {
+		// Check if client certificate is expired. If yes then delete the existing kubeconfig secret to make sure that the
+		// seed can be re-bootstrapped.
+		if err := shootClient.Client().Delete(ctx, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: gcc.KubeconfigSecret.Name, Namespace: gcc.KubeconfigSecret.Namespace}}); client.IgnoreNotFound(err) != nil {
+			return "", err
+		}
```
Comment (stoyanr) [inline pkg/gardenlet/controller/managedseed/managedseed_actuator.go]: I think you could use `kutil.DeleteSecretByReference` here?

File: pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go [hunk 8]
```diff
@@ -127,6 +129,25 @@ func (c *livecycleReconciler) Reconcile(ctx context.Context, req reconcile.Reque
...
+	if seed.Status.ClientCertificateExpirationTimestamp != nil && seed.Status.ClientCertificateExpirationTimestamp.UTC().Before(time.Now().UTC()) {
+		managedSeed := &seedmanagementv1alpha1.ManagedSeed{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      seed.Name,
+				Namespace: v1beta1constants.GardenNamespace,
+			},
+		}
+		if err := c.gardenClient.Client().Get(ctx, client.ObjectKeyFromObject(managedSeed), managedSeed); err != nil && !apierrors.IsNotFound(err) {
+			return reconcileResult(err)
+		} else if err == nil {
```
Comment (stoyanr) [inline pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go]: Perhaps this code could be extracted to an utility funciton, e.g. `kutil.GetManagedSeedByName` that would return nil if the managed seed doesn't exist? I think this is one of few places where such logic exists.

**Review by stoyanr — CHANGES_REQUESTED** [review]
Looks very good from my perspective, I have a few comments and suggestions mainly related to comments, consistency, and avoiding code duplication.
I would also like to test the change locally in a "remote garden" setup with a real managed seed.

File: pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go [hunk 9]
```diff
@@ -127,6 +129,25 @@ func (c *livecycleReconciler) Reconcile(ctx context.Context, req reconcile.Reque
...
+			metav1.SetMetaDataAnnotation(&managedSeed.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationReconcile)
```
Comment (stoyanr) [inline pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go:143]: Perhaps you could add a log message before this line (at least with `Debug` level)?

File: pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go [hunk 10]
```diff
@@ -127,6 +129,25 @@ func (c *livecycleReconciler) Reconcile(ctx context.Context, req reconcile.Reque
...
+			if err2 := c.gardenClient.Client().Status().Update(ctx, managedSeed); err2 != nil {
```
Comment (stoyanr) [inline pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go]: It's not the `Status` that we want to update here (and this breaks the entire mechanism because the managed seed is not updated). BTW, perhaps you could consider doing this update via a patch?
```suggestion
			if err2 := c.gardenClient.Client().Update(ctx, managedSeed); err2 != nil {
```

**Review by stoyanr — CHANGES_REQUESTED** [review]
I found an issue in the seed lifecycle controller while testing, see the comment below. When I fixed it as suggested below, the managed seed got re-reconciled and gardnlet was bootstrapped successfully.

**Comment by stoyanr** [comment]
While testing, I found also another issue. The requeue period for the seed lifecycle controller is 1 minute. However, gardenlet bootstrap sometimes may take longer than 1 minute. This causes the reconciliation of the managed seed to happen twice within 1 minute (below is a log message I added):

```
{"level":"info","msg":"Setting operation annotation for seed \"i500152-gcpx\" to reconcile.","seed":"i500152-gcpx","time":"2021-10-06T15:01:38+03:00"}
...
{"level":"info","msg":"Setting operation annotation for seed \"i500152-gcpx\" to reconcile.","seed":"i500152-gcpx","time":"2021-10-06T15:02:38+03:00"}
```

This may result in the `gardenlet-kubeconfig` secret being deleted during the second reconciliation, after `gardenlet` has already started, which results in the `gardenlet` logs being filled with errors.

**Edit:** I am not sure to what extent what I observed may be related to certain issues I had with my development setup at that time (that causes the bootstrap to take so much time). Once these issues have been resolved, I am no longer able to reproduce the above.

File: pkg/apis/core/types_seed.go (hunk 1 above)
Comment (rfranzke) [reply pkg/apis/core/types_seed.go]: I added the `Timestamp` suffix because of similarity to `.metadata.creationTimestamp` or `.metadata.deletionTimestamp`. It's a bit long, I agree, but for similarity reasons I'd keep it for now unless you insist on changing it.

File: pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_managedseed.go (hunk 4 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_managedseed.go:71]: When the `gardener.cloud/operation=reconcile` annotation is added to the `ManagedSeed` (e.g., by a human (manually) or by the seed lifecycle controller in GCM) then these checks would have prevented the `handleManagedSeedCreateOrUpdate` from being called.

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go (hunk 6 above)
Comment (rfranzke) [reply pkg/gardenlet/controller/managedseed/managedseed_actuator.go:694]: If a function's input parameters are each put into a dedicated line then - for symmetry reasons and to make it better readable - I typically also put the output parameters into a dedicated line.

File: pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go (hunk 10 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go]: Nice catch, thank you!

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go (hunk 6 above)
Comment (stoyanr) [reply pkg/gardenlet/controller/managedseed/managedseed_actuator.go:694]: OK, not insisting, I tend to split if the line is too long, but not otherwise, but obviously this is pure preference.

File: pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_managedseed.go (hunk 4 above)
Comment (stoyanr) [reply pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_managedseed.go:71]: OK, thanks!

File: pkg/apis/core/types_seed.go (hunk 1 above)
Comment (stoyanr) [reply pkg/apis/core/types_seed.go]: I don't insist of course, although I still think we should not feel obliged to add `Timestamp` to every `metav1.Time` field, especially when the resulting name is too long.

File: pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go (hunk 8 above)
Comment (rfranzke) [reply pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go]: I've done this, but not as part of `kutil`. I can move it to `kutil`, though, do you see other places of the code were using such function would be useful at this point in time?

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go [hunk 11]
```diff
@@ -344,13 +349,9 @@ func (a *actuator) deployGardenlet(
 	shootClient kubernetes.Interface,
 	managedSeed *seedmanagementv1alpha1.ManagedSeed,
 	gardenletConfig *configv1alpha1.GardenletConfiguration,
+	seed *gardencorev1beta1.Seed,
```
Comment (stoyanr) [inline pkg/gardenlet/controller/managedseed/managedseed_actuator.go:351]: Nit: Can we move this argument after `managedSeed` to achieve "perfect symmetry"?

**Review by stoyanr — COMMENTED** [review]
Thanks for addressing my feedback! Only a small nit, feel free to ignore it if you prefer. I will test this version before approving.

File: pkg/gardenlet/controller/managedseed/managedseed_actuator.go (hunk 11 above)
Comment (rfranzke) [reply pkg/gardenlet/controller/managedseed/managedseed_actuator.go:351]: Sure :) done!

File: pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go (hunk 8 above)
Comment (stoyanr) [reply pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go]: Yes, in the `handleManagedSeedIfSeedBelongsToIt` function that you added.

Comment (rfranzke) [reply pkg/controllermanager/controller/seed/seed_lifecycle_reconcile.go]: OK, thanks, done now in 8fc51127d2c311536dc4994c0cb602985b53a698.

File: pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_seed.go [hunk 12]
```diff
@@ -129,6 +128,13 @@ func (g *graph) handleSeedDelete(seed *gardencorev1beta1.Seed) {
 	g.deleteVertex(VertexTypeSeed, "", seed.Name)
 }
 
+func (g *graph) handleManagedSeedIfSeedBelongsToIt(ctx context.Context, seedName string) {
+	// error is ignored here since we cannot do anything meaningful with it
+	if managedSeed, err := kutil.GetManagedSeedByName(ctx, g.client, seedName); err != nil && managedSeed != nil {
```
Comment (stoyanr) [inline pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_seed.go]:
```suggestion
	if managedSeed, err := kutil.GetManagedSeedByName(ctx, g.client, seedName); err == nil && managedSeed != nil {
```
Or simply `managedSeed != nil` since if err is non-nil, managedSeed will always be nil.

**Review by stoyanr — CHANGES_REQUESTED** [review]

File: pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_seed.go (hunk 12 above)
Comment (rfranzke) [reply pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_seed.go]: Thanks...

**Review by stoyanr — APPROVED** [review]
/lgtm

Tested quickly with the latest changes as well, didn't encounter any new issues.

**Comment by timebertt** [comment]
/unassign
I won't make it to a review this week and @stoyanr already reviewed in-depth.
